pkg/agent/openflow/pipeline.go

```diff
@@ -606,181 +604,189 @@ func (c *featurePodConnectivity) conntrackFlows(category cookie.Category) []bind
+				NAT().
+				CTDone().
+				Done(),
+			ConntrackCommitTable.ofTable.BuildFlow(priorityHigh).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				MatchProtocol(ipProtocol).
+				MatchCTMark(ServiceCTMark).
+				Action().GotoStage(binding.OutputStage).
+				Done(),
+		)
+	}
+	// This generate the flow to match the first packet of Service connections.
+	targetTables := []uint8{SessionAffinityTable.ofTable.GetID(), ServiceLBTable.ofTable.GetID()}
+	if c.proxyAll {
+		targetTables = append([]uint8{NodePortProbeTable.ofTable.GetID()}, targetTables...)
 	}
+	flows = append(flows, ConntrackStateTable.ofTable.BuildFlow(priorityMiss).
+		Cookie(c.cookieAllocator.Request(category).Raw()).
+		Action().ResubmitToTables(targetTables...).
+		Done())
+	return flows
+}
 
-	// TODO: following flows should move to function "kubeProxyFlows". Since another PR(#1198) is trying
-	//  to polish the relevant logic, code refactoring is needed after that PR is merged.
-	for _, proto := range c.ipProtocols {
-		ctZone := CtZone
-		if proto == binding.ProtocolIPv6 {
-			ctZone = CtZoneV6
-		}
+// Feature: Service
+// Stage: PreRoutingStage
+// Tables: SNATConntrackTable
+// Stage: PostRoutingStage
+// Tables: SNATConntrackCommitTable
+// Refactored from:
+//   - part of func (c *client) connectionTrackFlows(category cookie.Category) []binding.Flow
+// snatConntrackFlows generates the flows related to SNAT ct zone.
+func (c *featureService) snatConntrackFlows(category cookie.Category) []binding.Flow {
+	var flows []binding.Flow
+	for _, ipProtocol := range c.ipProtocols {
 		flows = append(flows,
-			// Connections initiated through the gateway are marked with FromGatewayCTMark.
-			ConntrackCommitTable.BuildFlow(priorityNormal).MatchProtocol(proto).
+			// This flow is used to maintain SNAT conntrack for Service traffic.
+			SNATConntrackTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				MatchProtocol(ipProtocol).
+				Action().CT(false, SNATConntrackTable.ofTable.GetNext(), c.snatCtZones[ipProtocol]).
+				NAT().
+				CTDone().
+				Done(),
+			// This flow is used to mark the first packet of hairpin Service packet from status 'NotRequireSNATRegMark'
+			// to status 'CTMarkedSNATRegMark'.
+			SNATConntrackCommitTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				MatchProtocol(ipProtocol).
+				MatchCTStateNew(true).
+				MatchCTStateTrk(true).
+				MatchRegMark(HairpinRegMark).
+				MatchRegMark(NotRequireSNATRegMark).
+				Action().LoadRegMark(CTMarkedSNATRegMark).
+				Action().CT(true, SNATConntrackCommitTable.ofTable.GetID(), c.dnatCtZones[ipProtocol]).
+				LoadToCtMark(ServiceSNATCTMark).
+				CTDone().
+				Done(),
+			// This flow is used to mark the first packet of Service packet from Antrea gateway from 'RequireSNATRegMark'
+			// to status 'CTMarkedSNATRegMark'.
+			SNATConntrackCommitTable.ofTable.BuildFlow(priorityLow).
```

Comment:
Why this flow uses "priorityLow" but the packet with "NotRequireSNATRegMark" use "priorityNormal"?